### PR TITLE
MAINT: Use yarn as package manager

### PR DESCRIPTION
**Pull request by stevejpurves**
Using `yarn` in place of `npm` as the package manager

 - tested with `yarn` and resolved a number of issues around package dependencies.
 - updated contributing docs
 - added advisory around use of v7 and possible peer dependency errors
 - updates the `docs` build
 - updated `ci` actions
 - updates `dependabot`

Addresses #426

### CODE REVIEW

**Comment by stevejpurves** [comment]
Leaving this marked as WIP until ci goes green.

**Comment by stevejpurves** [comment]
ok, this is green and everything seems happy. I think this should make our CI checks more reliable 👍  and would love to get this in.

NOTE: if we accept this change, we should just delete all open dependabot PRs and wait on new ones coming in. Existing dependabot PRs are trying to patch `package-lock.json` which is now gone and we have updated dependencies with the PR anyways, because of the new `yarn-lock.json`

**Comment by moorepants** [comment]
Got these warnings on first install:

```
warning "@jupyter-widgets/base > @lumino/coreutils@1.8.0" has unmet peer dependency "crypto@1.0.1".
warning "@jupyterlab/theme-light-extension > @jupyterlab/application > @jupyterlab/ui-components@3.0.7" has unmet peer dependency "react@^17.0.1".
warning "@jupyterlab/testutils > ts-jest@25.5.1" has unmet peer dependency "typescript@>=3.4 <4.0".
warning " > jest-puppeteer@4.4.0" has incorrect peer dependency "puppeteer@>= 1.5.0 < 3".
warning " > ts-jest@27.0.4" has incorrect peer dependency "jest@^27.0.0".
warning " > ts-jest@27.0.4" has unmet peer dependency "typescript@>=3.8 <5.0".
```

File: .github/workflows/test.yml [hunk 1]
```diff
@@ -31,19 +31,19 @@ jobs:
...
-            ${{ runner.os }}-npm-
+            ${{ runner.os }}-yarn-
 
       - name: "Install"
         run: |
-          npm ci
+          yarn install --frozen-lockfile
 
-      - name: "`npm run fmt` and check for changes"
+      - name: "`nary run fmt` and check for changes"
```
Comment (moorepants) [inline .github/workflows/test.yml:42]:
```suggestion
      - name: "`yarn run fmt` and check for changes"
```

**Comment by moorepants** [comment]
When I run `yarn run test` locally jest isn't found. Should `yarn install` add jest? Or is there a separate lock file for dev dependencies?

**Comment by choldgraf** [comment]
Thanks for this PR - we should really get input from @minrk to make sure he is OK with this. (or @minrk feel free to say that you are fine with whatever @stevejpurves is proposing if you don't have the bandwidth to review etc!)

File: docs/contribute.md [hunk 2]
```diff
@@ -26,39 +26,63 @@ In order to get Thebe running locally, you'll need to have Node installed on you
...
+If you are using `npm` v7 you may encounter **Peer Dependency Errors** see comments below to resolve these.
+
 ```{note}
-Using `npm ci` rather than `npm install` will ensure that you install the latest tested dependencies, and will not make any unintentional local upgrades.
+Using `yarn install` will ensure that you install the latest tested dependencies, and will not make any unintentional local upgrades. `yarn` uses `npm` under the hood, please do not use `npm install` directly to install dependencies.
+```
+
+#### Peer Dependency Errors
+
+Due to recent changes in `npm` as of v7 peer dependency issues are flagged as critical errors. However, may projects have not yet resolves these issues in their code base, this means that a project's co-dependencies can prevent installation of a package. This is a know issue, to resolve this there are two options:
```
Comment (moorepants) [inline docs/contribute.md:66]:
```suggestion
Due to recent changes in `npm` as of v7 peer dependency issues are flagged as critical errors. However, many projects have not yet resolved these issues in their code base, this means that a project's co-dependencies can prevent installation of a package. This is a know issue, to resolve this there are two options:
```

**Comment by moorepants** [comment]
There is an empty `yarn-lock.json` file committed along with the two `yarn.lock` files. Maybe that is unintentional?

**Comment by moorepants** [comment]
This looks good to me so far.

1. I'm not sure how to install the full development dependencies to run the tests.
2. We'll likely have to check the production build commands on master when we do a release for a full check of the CI.

Thanks @stevejpurves !

**Comment by stevejpurves** [comment]
@moorepants to your points above

 - yarn-lock.json removed and the docs makefile is now fixed
 - I have addressed a couple of them. I'm tempted to say leave the others, though? (react, typescript) these are our dependencies telling us we are responsible for providing these peer dependencies, but looking at each warning there, I don't think we are using our dependencies in way where we need to install react, typescript, so would be best not to add them.

**Comment by moorepants** [comment]
I'll try again.

**Comment by moorepants** [comment]
Everything is working now. Not sure what I did wrong before, maybe I didn't delete the `node_modules` directory first.

**Comment by moorepants** [comment]
+1 from me on this for merging.

**Review by choldgraf — APPROVED** [review]
I also think this is a nice enhancement! Don't mistake my comments above for a lack of enthusiasm!

**Comment by stevejpurves** [comment]
no worries @choldgraf it's a good point that we're all on different cadences here and need to allow some time for wider contributions :)

**Comment by stevejpurves** [comment]
Are we good to merge this one folks? @moorepants @choldgraf (I have no merge button :) )

**Comment by choldgraf** [comment]
Sounds good! Thanks for the ping 🙂

**Pull request merged**